uefitoolPackages.new-engine: A62 -> A71

athre0z opened the pull request:
###### Description of changes

This commit upgrades uefiToolPackages.new-engine to the latest version as of writing. Version A68 is now using Qt6 and no longer comes with a custom build script, so the build logic compared to the old-engine variant has become sufficiently divergent that I decided that it would be easier to split them into two entirely different derivations.

Additionally, I:

- added myself as a maintainer
- added a patch that allows the new-engine variant to build on Darwin
- correctly set `mainProgram` in the old-engine variant

A68 release on GitHub: https://github.com/LongSoft/UEFITool/releases/tag/A71

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux (via qemu)
  - [ ] x86_64-darwin (via rosetta)
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ajs124 commented:
is the old-engine variant still useful? if not, maybe we can just drop it.

athre0z commented:
Theoretically the `A` in the version number of the new-engine variant stands for `alpha`, but on the other hand the old engine hasn't [seen an update in 3 years](https://github.com/LongSoft/UEFITool/releases/tag/0.28.0). Not sure if people are still using it. It's not really much of a hassle to keep it, so I figured I'd leave it in until the devs condider the new engine to no longer be "alpha".

jiegec commented:
Result of `nixpkgs-review pr 240420` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>uefitool (uefitoolPackages.new-engine)</li>
  </ul>
</details>

jiegec commented:
Result of `nixpkgs-review pr 240420` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>uefitool (uefitoolPackages.new-engine)</li>
  </ul>
</details>

athre0z commented:
Rebased on master -- no other changes.

athre0z commented:
Rebased on master and updated to [A68](https://github.com/LongSoft/UEFITool/releases/tag/A68).

athre0z commented:
Rebased on master.

athre0z commented:
Happy to rebase this if it helps. I had sort of abandoned it after it hadn't been reviewed / approved for more than a year.

lopter commented:
> Happy to rebase this if it helps. I had sort of abandoned it after it hadn't been reviewed / approved for more than a year.

I guess it would be nice to update the PR with the rebase I did, maybe you can get a fresh checkout of nixpkgs, apply the [diff] from the commit I made earlier, and update the PR?

Anyway, thank you for the work, I am trying to help with some old Thinkpad, and it needs Coreboot in order to make some hardware upgrade, and apparently uefitool is needed in the process.

[diff]: https://github.com/NixOS/nixpkgs/commit/84e5df5086adb9ef5169a457c691bd9226ecf3b4.diff

athre0z committed:
uefitoolPackages.new-engine: A62 -> A71

This commit upgrades uefiToolPackages.new-engine to the latest version
as of writing. Version A68 is now using Qt6 and no longer comes with
a custom build script, so the build logic compared to the old-engine
variant has become sufficiently divergent that I decided that it would
be easier to split them into two entirely different derivations.

Additionally, I:

- added myself as a maintainer
- added a patch that allows the new-engine variant to build on Darwin
- correctly set `mainProgram` in the old-engine variant

athre0z commented:
Thanks for providing the updated diff: did as you suggested! :)

zimward commented:
whats the status of this? also in the mean time upstream released version `A71`

athre0z commented:
Happy to update it, though at this point I don't exactly have high hopes of it being reviewed & merged. Not sure how to proceed here.

zimward commented:
if the conflict gets resolved and you update it i am sure that we can get it merged in the not too soon future.

athre0z commented:
Rebased and updated to A71. Let's see if we can get it merged this time.

zimward commented on pkgs/tools/system/uefitool/new-engine.nix:
```diff
@@ -0,0 +1,37 @@
+  ];
+  patches = lib.optionals stdenv.isDarwin [ ./bundle-destination.patch ];
+
+  meta = with lib; {
+    description = "UEFI firmware image viewer and editor";
+    homepage = "https://github.com/LongSoft/uefitool";
+    license = licenses.bsd2;
+    maintainers = with maintainers; [ athre0z ];
+    platforms = platforms.linux ++ platforms.darwin;
+  };
```
```suggestion
  meta = {
    description = "UEFI firmware image viewer and editor";
    homepage = "https://github.com/LongSoft/uefitool";
    license = lib.licenses.bsd2;
    maintainers = with lib.maintainers; [ athre0z ];
    platforms = lib.platforms.linux ++ lib.platforms.darwin;
    mainProgram = "UEFITool"
  };
```
The usage of `with lib;` outside of small scopes has been considered as an anti-pattern fore some time now in `nixpkgs`. See #208242

zimward commented on pkgs/tools/system/uefitool/old-engine.nix:
```diff
@@ -43,15 +33,16 @@ mkDerivation rec {
+    cp UEFITool UEFIReplace/UEFIReplace UEFIPatch/UEFIPatch "$out"/bin
   '';
 
   meta = with lib; {
     description = "UEFI firmware image viewer and editor";
     homepage = "https://github.com/LongSoft/uefitool";
     license = licenses.bsd2;
```
```suggestion
  meta = {
    description = "UEFI firmware image viewer and editor";
    homepage = "https://github.com/LongSoft/uefitool";
    license = lib.licenses.bsd2;
```

zimward commented on pkgs/tools/system/uefitool/old-engine.nix:
```diff
@@ -43,15 +33,16 @@ mkDerivation rec {
     homepage = "https://github.com/LongSoft/uefitool";
     license = licenses.bsd2;
-    maintainers = [ ];
+    maintainers = with maintainers; [ athre0z ];
     # uefitool supposedly works on other platforms, but their build script only works on linux in nixpkgs
     platforms = platforms.linux;
```
```suggestion
    maintainers = with lib.maintainers; [ athre0z ];
    # uefitool supposedly works on other platforms, but their build script only works on linux in nixpkgs
    platforms = lib.platforms.linux;
```

zimward commented on pkgs/tools/system/uefitool/new-engine.nix:
```diff
@@ -0,0 +1,37 @@
+  src = fetchFromGitHub {
+    sha256 = "sha256-NRlrKm5+eED6oyvFRSEhn0EUbMsPJtuFAyv3vgY/IUI=";
+    owner = "LongSoft";
+    repo = pname;
+    rev = version;
```
```suggestion
    repo = "uefitool";
    tag = version;
```
the repo shouldn't change when overriding the `pname` and a few months ago the `tag` attr was added to the gh fetcher (see #355973 ).

zimward reviewed (changes requested):
A few conventions changed since you first opened the PR, after that i think it should be in a pretty good state.

athre0z commented:
Thank you! I addressed all feedback. I slightly adjusted your suggested `meta.mainProgram` to `"uefitool"` since the executable name has changed in the new engine variant.

zimward reviewed (approved):
LGTM

LordGrimmauld commented on pkgs/tools/system/uefitool/new-engine.nix:
```diff
@@ -0,0 +1,38 @@
+  zip,
+}:
+
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

rec is discouraged/deprecated in nixpkgs, the finalAttrs pattern is the current convention.

athre0z replied:
I applied this for `new-engine.nix` as suggested, but left `rec` for `old-engine.nix`. The latter gets the `mkDerivation` filled in from `libsForQt5.callPackage` which appears to not support the lambda / `finalAttrs` variant of the invocation. I can use a `let` instead, if that is preferred over using `rec`?

LordGrimmauld replied:
Don't use `let ... in`, that one is terrible for overrides. `rec` is still fine if finalAttrs is unavailable.

LordGrimmauld commented on pkgs/tools/system/uefitool/variants.nix:
```diff
@@ -1,24 +1,5 @@
-  };
+{ libsForQt5, qt6Packages }:
+{
+  new-engine = qt6Packages.callPackage (import ./new-engine.nix) { };
+  old-engine = libsForQt5.callPackage (import ./old-engine.nix) { };
```
```suggestion
  new-engine = qt6Packages.callPackage ./new-engine.nix { };
  old-engine = libsForQt5.callPackage ./old-engine.nix { };
```
I believe callPackage can consume paths directly

LordGrimmauld commented on pkgs/tools/system/uefitool/old-engine.nix:
```diff
@@ -14,19 +9,14 @@
 
   src = fetchFromGitHub {
-    inherit sha256;
+    sha256 = "1n2hd2dysi5bv2iyq40phh1jxc48gdwzs414vfbxvcharcwapnja";
```
```suggestion
    hash = "...";
```
this will still be a sha hash, but FOD hashes are supposed to be defined via `hash` now.

LordGrimmauld commented on pkgs/tools/system/uefitool/new-engine.nix:
```diff
@@ -0,0 +1,38 @@
+    homepage = "https://github.com/LongSoft/uefitool";
+    license = lib.licenses.bsd2;
+    maintainers = with lib.maintainers; [ athre0z ];
+    platforms = lib.platforms.linux ++ lib.platforms.darwin;
```
```suggestion
    platforms = lib.platforms.unix;
```
Might `unix` make more sense here? in practice that means darwin+linux as that is what our CI covers. However, it doesn't straight disregard the possibility of running this on e.g. bsd.

LordGrimmauld reviewed (changes requested):
Have some nits.
Packages do build on `x86_64-linux`. Cross is broken because qt cross is a mess right now, can't really test the package itself. Building against musl and LLVM right now, but that'll take a few hours and is probably not blocking either.

LordGrimmauld reviewed (approved):
Diff looks good and the thing builds, though i do not feel competent judging the functionality of the actual software

zimward commented:
functionality wise it looks fine. i inspected some bios images and everything looks like i would expect. (editing is sadly still not supported in NE, so not much more to test)
